### PR TITLE
feat: new DocTypes "Code List" and "Common Code"

### PR DESCRIPTION
**Pull request by barredterra**
code_list

common_code_list

common_code

```python
In [1]: code_list = "urn:xoev-de:kosit:codeliste:untdid.5305_3"

In [2]: get_codes_for(code_list, "Account", "Tax Assets - WPL")
Out[2]: ('A',)

In [3]: get_docnames_for(code_list, "Account", "A")
Out[3]: ('Tax Assets - WPL',)

In [4]: get_default_code(code_list)
Out[4]: 'S'
```

Screenshots based on https://www.xrepository.de/api/xrepository/urn:xoev-de:kosit:codeliste:untdid.5305_3:technischerBestandteilGenericode:datei:UNTDID_5305-3.xml

After this is merged, check if we should revert https://github.com/frappe/erpnext/pull/40593 and https://github.com/frappe/erpnext/pull/40663

> no-docs (for now)

### CODE REVIEW

**Comment by blaggacao** [comment]
- [x] added fully tested genericode importer
- [x] added wip of an edi workspace
- [ ] make it hidden by default (check it actually works)
- [x] added wip of edi template
- [ ] dogfood and validate snippets with current (operational) implementation of colombia e-fact
- [ ] try out jpype and use java libraries for validation
- [ ] try to implement a wsdl endpoint per edi_template (hence `get_parsed`)

<details><summary>Wsdl service skeletton</summary>

```python
from spyne import Application, rpc, ServiceBase, Unicode
from spyne.protocol.soap import Soap11
from spyne.server.wsgi import WsgiApplication

class HelloWorldService(ServiceBase):
    @rpc(Unicode, _returns=Unicode)
    def say_hello(ctx, name):
        return f"Hello, {name}!"

application = Application([HelloWorldService],
    tns='http://example.com/hello',
    in_protocol=Soap11(validator='lxml'),
    out_protocol=Soap11())

# Generate WSDL
from spyne.interface.wsdl import Wsdl11
wsdl_content = Wsdl11(application.interface).get_interface_document()

# Save WSDL to file
with open('hello_service.wsdl', 'wb') as f:
    f.write(wsdl_content)
```

</details>

**Comment by barredterra** [comment]
- [x] It is not obvious (and was not intended by me) that the `CanonicalVersionUri` now has to be used as the name of the code list (for the import to work). E.g. I would not expect the `CanonicalVersionUri` of the code list linked above to be referenced in most eInvoices. So a more generic name would be useful for retrieving the necessary code list. In the screenshot above, I used the `Identification/ShortName`.
- [x] "Additional Data" from the code list is not yet imported (e.g. descriptions, as available in the code list linked above).

**Comment by blaggacao** [comment]
More exploration from my side: https://github.com/barredterra/erpnext/pull/1 (Edi Template & Edi Log, squashed your naming fix 
there, @barredterra )

**Comment by blaggacao** [comment]
> It is not obvious (and was not intended by me) that the CanonicalVersionUri now has to be used as the name of the code list (for the import to work). E.g. I would not expect the CanonicalVersionUri of the code list linked above to be referenced in most eInvoices.

**Firstly**, most of the time, however, unlike the common codes, these lists are akin to real _constants_ in an e-invoice template. I would typically hard-code them for each distinct edi template. Here is an example:

<details><summary>XML Example</summary>

```xml
        <cbc:IdentificationCode
          listAgencyID="6"
          listAgencyName="United Nations Economic Commission for Europe"
          listSchemeURI="urn:oasis:names:specification:ubl:codelist:gc:CountryIdentificationCode-2.1"
        >CO</cbc:IdentificationCode>
```

</details>

**Secondly**, the title is still user friendly, for display purposes.

<details><summary>Screenshot</summary>

image

</details>

**Thirdly**, as per the Genericode XML Schema, `CanonicalVersionUri` is the primary key of a list, so we would have to go out of our way to work around that. https://docs.oasis-open.org/codelist/genericode/v1.0/os/xsd/genericode.xsd

<details><summary>XML Reference</summary>

```xml
<xsd:complexType name="Identification">
<xsd:annotation>
<xsd:documentation>Identification and location information (metadata).</xsd:documentation>
</xsd:annotation>
<xsd:sequence>
<xsd:group ref="gc:NameSet">
...
</xsd:group>
<xsd:element name="Version" type="xsd:token">
...
</xsd:element>
<xsd:element name="CanonicalUri" type="xsd:anyURI">
<xsd:annotation>
<xsd:documentation>Canonical URI which uniquely identifies all versions (collectively).</xsd:documentation>
<xsd:documentation>
<rule:text id="R25" category="document">Must be an absolute URI, must not be relative.</rule:text>
</xsd:documentation>
<xsd:documentation>
<rule:text id="R26" category="application">Must not be used as a de facto location URI.</rule:text>
</xsd:documentation>
</xsd:annotation>
</xsd:element>
<xsd:group ref="gc:VersionLocationUriSet">
...
</xsd:group>
<xsd:element name="Agency" minOccurs="0" type="gc:Agency">
...
</xsd:element>
</xsd:sequence>
</xsd:complexType>

...

<xsd:group name="VersionLocationUriSet">
<xsd:annotation>
<xsd:documentation>Identification and location URIs for a version.</xsd:documentation>
</xsd:annotation>
<xsd:sequence>
<xsd:element name="CanonicalVersionUri" type="xsd:anyURI">
...
</xsd:element>
<xsd:element name="LocationUri" minOccurs="0" maxOccurs="unbounded" type="xsd:anyURI">
...
</xsd:element>
<xsd:element name="AlternateFormatLocationUri" minOccurs="0" maxOccurs="unbounded" type="gc:MimeTypedUri">
...
</xsd:element>
</xsd:sequence>
</xsd:group>
```

Note, that `CanonicalUri` _will_ be shared between lists, and within `VerstionLocationUriSet` (the unique identifying set), `CanoncialVersionUri` is the only mandatory one.

</details>

**Edit:**

I think this is what you where looking for:

image

**Review by blaggacao — DISMISSED** [review]
I have contributed the Genericode Importer which triggered a certain amount of polishing and I think this version is ready to merge, now.

We should keep in mind, that we're actively developing the EDI functionality for EU & Colombia in an iterative approach  against `develop` - so more to come, soon.

**Comment by barredterra** [comment]
If we have to use this very technical name, we should at least offer the import from the list view (or from an unsaved form) instead of from the saved form view. It would be kind of weird if the user first needs to manually extract some values from the XML, in order to then import the rest of it.

My approach was originally targeted at, e.g., being able to tell the users to create a **Code List** called "UNTDID.5305" and make sure it contains the tax codes they care about. The alternative, equally good or even better, would be being able to tell them to just import [this code list](xrepository.de/api/xrepository/urn:xoev-de:kosit:codeliste:untdid.5305_3:technischerBestandteilGenericode:datei:UNTDID_5305-3.xml). Telling them to extract the `CanonicalVersionUri` does not seem realistic, however.

**Comment by blaggacao** [comment]
> [ ... ] we should at least offer the import from the list view (or from an unsaved form) instead of from the saved form view. It would be kind of weird if the user first needs to manually extract some values from the XML, in order to then import the rest of it.

> Telling them to extract the CanonicalVersionUri does not seem realistic, however.

Indeed! &mdash; I think this concern is already gracefully implemented?

### General Scenario: Add more codes from the same list (e.g. different filters) &mdash; form import
#### Specific Scenario: I accidentially selected the wrong code list for upload
image

---

### General Scenario: Add a new code list and some codes &mdash; list import
Implemented on the list views of _Code List_ and _Common Code_
#### Specific Scenario: Step Two (immediately after upload); Selection of Import filters
image
<sup>Note: no extraction of `CanonicalVersionUri` by the user; filterable are fields with cardinality <= 5</sup>

**Comment by blaggacao** [comment]
https://github.com/frappe/erpnext/pull/43425/commits/d540d03d8a5903764c311b8d19c3d7cb3188e6d9

image

**Comment by blaggacao** [comment]
We may need to relax the unique condition on `code` for _Common Code_. This is a valid Code List:
```xml
      <Row>
         <Value ColumnRef="code">
            <SimpleValue>1.50</SimpleValue>
         </Value>
         <Value ColumnRef="name">
            <SimpleValue>ReteFuente</SimpleValue>
         </Value>
         <Value ColumnRef="description">
            <SimpleValue>Compras con tarjeta débito o crédito</SimpleValue>
         </Value>
      </Row>
      <Row>
         <Value ColumnRef="code">
            <SimpleValue>1.50</SimpleValue>
         </Value>
         <Value ColumnRef="name">
            <SimpleValue>ReteFuente</SimpleValue>
         </Value>
         <Value ColumnRef="description">
            <SimpleValue>Compras de bienes o productos agrícolas o pecuarios sin procesamiento industrial</SimpleValue>
         </Value>
      </Row>
```

**Comment by blaggacao** [comment]
Problem addressed, also in the importer:

image

image

**Comment by blaggacao** [comment]
https://github.com/frappe/erpnext/pull/43425/commits/52dcb2bca220aa4def00e5a338fb11b60943fe02 addresses the usability concern even further:

image

**Comment by blaggacao** [comment]
## Testing Resume

- [x] 23 Code Lists, 27 XML files
- [x] Faulty XML: ok
- [x] Undeclared Columns: ok
- [x] Non unique `common_code`: ok (eg tax rates) -> added description
- [x] Guard against accidential structured data manipulation
- [x] Partial loading of a list
- [x] Later completion of a list - USER ATTENTION! Duplicates are legal in the data structure, double loading must be avoided with care (no other way)
- [x] Loading values from different files into the same code list 

<details><summary>Screenshot</summary>

image

</details>

**Comment by barredterra** [comment]
> Indeed! — I think this concern is already gracefully implemented?

Ah, my bad, I've overlooked this.

---

I think, from a developer perspective, it would be perfect if we can:

(1) Given a ref_doctype, ref_docname and code list, retrieve the applicable code.
(2) Given a code and code list, retrieve the applicable ref_doctype and ref_docname.

Do you think this unambiguous mapping is feasible?

I'm thinking of adding a util and validation for the second case.

---

Selecting "URL" on file upload fails. Would be nice if we can support this as well.

---

File: erpnext/edi/doctype/code_list/code_list_import.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+		content=frappe.local.uploaded_file,
+		dt="Code List",
+		dn=code_list.name,
+		folder="Home/Attachments",
+		is_private=1,
+	)
+	file_path = file_doc.get_full_path()
+
+	# Get available columns and example values
+	columns, example_values, filterable_columns = get_columns_and_examples(file_path)
```
Comment (barredterra) [inline erpnext/edi/doctype/code_list/code_list_import.py]:
```suggestion
	columns, example_values, filterable_columns = get_columns_and_examples(root)
```

I think we can pass the already parsed XML here.

**Comment by blaggacao** [comment]
> (2) Given a code and code list, retrieve the applicable ref_doctype and ref_docname.

<details><summary>Given ...</summary>

```python

	def validate_distinct_references(self):
		"""Ensure no two Common Codes of the same Code List are linked to the same document."""
		for link in self.applies_to:
			existing_links = frappe.get_all(
				"Common Code",
				filters=[
					["name", "!=", self.name],
					["code_list", "=", self.code_list],
					["Dynamic Link", "link_doctype", "=", link.link_doctype],
					["Dynamic Link", "link_name", "=", link.link_name],
				],
				fields=["name", "common_code"],
			)

			if existing_links:
				existing_link = existing_links[0]
				frappe.throw(
					_("{0} {1} is already linked to Common Code {2}.").format(
						link.link_doctype,
						link.link_name,
						get_link_to_form("Common Code", existing_link["name"], existing_link["common_code"]),
					)
				)
```

</details>

I think we can restrict certain Code Lists (optionally) to a certain doctype and thereby meet the full condition necessary to implement such a util for such code lists.

But you're absolutely right: on the parsing end of the flow, this would quite definitely improve DX.

**Comment by blaggacao** [comment]
> Selecting "URL" on file upload fails. Would be nice if we can support this as well.

https://github.com/frappe/erpnext/pull/43425/commits/94bd74319ba8f5641c1ff5dc5e192fe296bd0113

Also library links (for example to post-load more codes from an already partially imported list)

Depends on: https://github.com/frappe/frappe/pull/27954

**Comment by blaggacao** [comment]
We probably should also return `additional_data` from the `get_code_for` because more often than not, this is actually relevant data when rendering a e-invoice template. I'll check this and confirm during dogfooding.

---

Wile dogfooding I realise:

As user, I want an easy way to associate all Code List documents with my current document, which are configured to be associatable with that document type (by means of a Code List - Doctype association). Especially for documents which are still metadata, but fall within the life-cycle management of non-power users, such as Suppliers, Customers & Items.

---

We also should probably implement a stringer (`__str__`) on `CommonCode` which automatically renders it's `common_code` into a jinja template without need for more verbosity.

---

`get_code_for` should probably throw a user facing error if no association is found but was expected, indicating that xyz association is missing and needs to be set up.

```python
    # should throw if tax_category_id was not set
    tax_category_id = frappe.get_doc("Code List", TAX_CATEGORY_CODE_LIST).get_code_for(
        party.doctype, party.name
    )
```

---

_Code List_ should have a default code, so that it is easier to fetch:

```python
    tax_id_type = frappe.get_doc("Common Code", party.tax_id_type or tax_type_codelist.default)
```

---

Would be handy too (`.get_doc` proxy):
```python
    tax_id_type = codelist.get_doc(party.tax_id_type)  # raises
    tax_id_type = codelist.get_doc(party.tax_id_type, None)
    tax_id_type = codelist.get_doc(party.tax_id_type, codelist.default)
```

---

Ultimately, what I want is this, to keep my implementation at the leaf as clean as possible:
```python
from .code_list import substitute


customer = frappe.get_doc("Customer", "Foo")

customer_for_render = substitute(customer)
# replaces all links to Common Code to the respective common code
# and infers defaults based on the field's filter
# {"code_list": "My List"} - take default of that list, if there is
```

<sup> If my edge case invalidates a default, I can still override it in sitiu </sup>

**Comment by blaggacao** [comment]
@barredterra I noted a couple of things, when dogfooding this yesterday.

1. On "intermediate" metadata (like Customer, Supplier, which are metadata, but where the lifecycle is managed as part of daily transactions), I think we still want a "nomarl" link from a custom field to Common Code with the appropriate filter.
2. On these cases it would be wonderful to have a summary `customer = code_list.substitute(customer)` to replace all these link fields with a Common Code object for use during rendering. (benefits: same mental model as the origin doctype, just augmented, avoids running into the trap of creating own `frappe._dict()` ad-hoc data structures)
3. Because then, we get into Custom Field management already, I think it would be fair to add a (rudimentary) wizard of some sort to the Code List doctype which allows to setup such custom fields &mdash; this would be an alternative mechanism of association for the use case of those "indermediate" metadata in addition to the Common Code's `applies_to` for "long lived" metadata.

This doesn't mean the `applies_to` is useless, it's still extremely valid on quite an array of very stable metadata, but I think for the transaction-bound metadata (read: Customer), we need an in-transaction complement.

You may consider my previous list as "dogfooding notes", I haven't fully processed them myself, yet. But it seems that my refactoring would be stuck on what I describe here, for the "true blessing of DX".

**Comment by blaggacao** [comment]
> (2) Given a code and code list, retrieve the applicable ref_doctype and ref_docname.

Since this happens in the context of parsing, and parsing is bound to a XML structure, the following binding may be relevant, as well:

image

**Meaning**
- In the context of DIAN E-Invoice
- With its binding to a Sales Invoice
- And the associated code lists bound to python context variables `pais` & `to` ("given code list")
- Then infer the code's `refdoc_doctype` & `refdoc_name` by means of a triangulation with an additional Doctype Context binding in the `applies_to` table.

Originally, I had thought of an implementation with a direct Code List - Doctype binding, but I realize that the EDI Template is a better source for that contextual binding. Hence having such a context binding within the `applies_to` table might make more sense.

It would read in plain english:

1. In the context of Sales Invoice EDI
2. Apply the present code
3. To the target document (e.g. an _Account_ / Tax) 

Now, instead of binding it to a Doctype, we probably could bind it to an EDI Template within `applies_to` to make true with the above plain english reading of the feature.

**Comment by blaggacao** [comment]
... which makes me wonder: probably we should bind all `applies_to` inside the EDI template (as in: "Within the context of a particular EDI type")? :thinking: 

---

image

**Comment by blaggacao** [comment]
https://github.com/frappe/erpnext/pull/43425/commits/d68d80892db14b62e47f60a22b441fa62e5b77b6

Fix UX of Title-less common codes in link dropdowns:

image

**Comment by blaggacao** [comment]
Conclusion: We need to incubate a core feature "LazyRecords", which lazily traverse link values on on access but eagerly build the traversal tree 3-4 levels deep for best autocompletion. Inside, we then can replace standard Link bindings with our custom bindings. Whereby barren key renders the stringer of the target class in jinja (typically: name or title(?)). :rofl:

**Comment by barredterra** [comment]
https://github.com/frappe/erpnext/pull/43425#issuecomment-2385467354

I broke this again, sorry.

Is this really valid? What use is a code if it can have multiple meanings, even in the same list?

**Comment by blaggacao** [comment]
> Is this really valid?

Yes.

> What use is a code if it can have multiple meanings, even in the same list?

Not sure if we productively can second-guess the spec &mdash; especially since I have a real codelist which exhibits this behaviour.

Remember that `code` is not guaranteed to be a row prim key (there isn't even a guarantee that a row prim key is present), so we can't fake it and make it a requirement in our implementation.

File: erpnext/edi/doctype/common_code/common_code.py [hunk 2]
```diff
@@ -60,52 +59,68 @@ def validate_distinct_references(self):
...
+
+		if frappe.db.exists("Common Code", code_name):
+			common_code = frappe.get_doc("Common Code", code_name)
+		else:
+			common_code: "CommonCode" = frappe.new_doc("Common Code")
+			common_code.common_code = code_value
+			common_code.code_list = code_list
+
+		common_code.from_genericode(column_map, xml_element)
+		common_code.save()
```
Comment (blaggacao) [inline erpnext/edi/doctype/common_code/common_code.py:107]: This is unfortunately prohibitively slow :cry: &mdash; we should revert back to the bulk insert.

I don't quite see a use case to update a code list item: they don't change on the same version. If they change they change on a different version which is a different code list.

File: erpnext/edi/doctype/common_code/common_code.py [hunk 3]
```diff
@@ -60,52 +59,68 @@ def validate_distinct_references(self):
...
+	filter_conditions = [
+		f"Value[@ColumnRef='{column_ref}']/SimpleValue='{value}'" for column_ref, value in filters.items()
+	]
+	if filter_conditions:
+		xpath_expr += "[" + " and ".join(filter_conditions) + "]"
+
+	for xml_element in root.xpath(xpath_expr):
+		code_column = column_map["code"]
+		code_value = xml_element.find(f"./Value[@ColumnRef='{code_column}']/SimpleValue").text
+		code_name = get_code_name(code_list, code_value)
```
Comment (blaggacao) [inline erpnext/edi/doctype/common_code/common_code.py]: `idx` is required here to prevent name conflicts (make them less likely \*). The hashing is still valuable as it provides predictable / independently reproducible names.

\* They can still occur when importing from two files where exactly at the same `idx` (after filtering) we have the same `code` &mdash; we should fix that better, actually.

**Review by blaggacao — DISMISSED** [review]
I appreciate the better codelist handling &mdash; that looks a lot cleaner.

We'd just need to mitigate the two regressions:

- Insertion speed for Codes (removal of raw bulk insert) &mdash; I think this can be still achived after the clarifying refactorings by operating on a generator / list rather than on the individual record object
- Unique naming &mdash; we should aim to fix the silent bug in the previous implementation, as well

> I broke this again, sorry.

Maybe I overlooked. But I'm not even quite sure if you broke it in principle :smile: or just accidentially via removing the `idx` during import. :smile:

File: erpnext/edi/doctype/code_list/code_list.py [hunk 4]
```diff
@@ -78,3 +82,23 @@ def get_code_for(code_list: str, doctype: str, name: str) -> str | None:
...
+		.on((CommonCode.name == DynamicLink.parent) & (DynamicLink.parenttype == "Common Code"))
+		.select(DynamicLink.link_name)
+		.where(
+			(DynamicLink.link_doctype == doctype)
+			& (CommonCode.common_code == code)
+			& (CommonCode.code_list == code_list)
+		)
+	).run()
+
+	return docnames[0][0] if docnames else None
```
Comment (blaggacao) [inline erpnext/edi/doctype/code_list/code_list.py]: So, I guess this will break in a subset of valid cases, namely when we can't assume `common_code` to be a unique column, right?

I think we should avoid to offer unbound behaviour. For example, we could return a UserError condition if we detect more than one row in the query result. This would even relax the unique constraint required by this getter to `(codelist, doctype, code)` without precluding that the underlying dataset can't exhibit such records.

File: erpnext/edi/doctype/common_code/common_code.py (hunk 2 above)
Comment (barredterra) [reply erpnext/edi/doctype/common_code/common_code.py:107]: Prohibitively slow? I could't even blink while this imported my test data. How big are your code lists and how many do you want to import?

To be a bit more precise, on my machine this has a throughput of >200 records per second. And it needs to be done only once.

File: erpnext/edi/doctype/common_code/common_code.py (hunk 3 above)
Comment (barredterra) [reply erpnext/edi/doctype/common_code/common_code.py]: So let's just use normal hash naming? I don't see the need for predictable / independently reproducible names and using idx seems kind of random (except the order is guaranteed within the code list).

File: erpnext/edi/doctype/common_code/common_code.py (hunk 2 above)
Comment (blaggacao) [reply erpnext/edi/doctype/common_code/common_code.py:107]: > Prohibitively slow?

It was at least pretty annoying during local tests, so much so that I took care to make it faster (not by accident or not knowing better). Now you took the same care to make it slower (for other reasons not entirely evident to even though I tried to infer some), so I'll have to assume it's going to be "pretty annoying during local tests" again.

Maybe you want to first briefly explain these "other reasons"?

File: erpnext/edi/doctype/common_code/common_code.py (hunk 3 above)
Comment (blaggacao) [reply erpnext/edi/doctype/common_code/common_code.py]: > So let's just use normal hash naming?

Yeah, that would solve the issue for good. :+1:

File: erpnext/edi/doctype/common_code/common_code.py (hunk 2 above)
Comment (barredterra) [reply erpnext/edi/doctype/common_code/common_code.py:107]: I usually avoid `bulk_insert`. Yes, it is fast, but the reason for that is that it skips all validations and hooks, making it error prone and hard to customize. I would rather wait a couple seconds instead of skipping all that good stuff. Sure, when a task would take many minutes or hours, we cannot avoid bulk_insert. But for saving a couple of seconds, I think it's not worth the (potential) trouble.

Comment (barredterra) [reply erpnext/edi/doctype/common_code/common_code.py:107]: I have now added a progress bar, so we don't have to stare at a blank screen during longer imports.

Comment (blaggacao) [reply erpnext/edi/doctype/common_code/common_code.py:107]: I see. Thanks. That makes sense, in principle.

Do we realistically expect the need for validations and hooks on _mass_ import of _Common Codes_ in practice, though? &mdash; I have a bit of a hard time to come to terms with the ask and presumption to deprioritize speed on abstract principle, alone.

The scenario here simply seems a bit far fetched to me.

The progress bar is a good idea to mitigate the problem. At least people wouldn't try to reload the page because it goes unresponsive.

It's been a while and I can't remember exactly how long it took. Granted, "prohibitively", indeed says as much about my personal tolerance level for slowness. 😊 Let me check it again report back, and when it takes more than 5 seconds, we revert back?

**Comment by barredterra** [comment]
For testing, here's a list of code lists required for eInvoicing in Germany (XRechnung):

- https://www.xrepository.de/api/xrepository/urn:xoev-de:kosit:codeliste:rec21_3:technischerBestandteilGenericode
- https://www.xrepository.de/api/xrepository/urn:xoev-de:kosit:codeliste:rec20_3:technischerBestandteilGenericode
- https://www.xrepository.de/api/xrepository/urn:xoev-de:kosit:codeliste:vatex_1:technischerBestandteilGenericode
- https://www.xrepository.de/api/xrepository/urn:xoev-de:kosit:codeliste:eas_5:technischerBestandteilGenericode
- https://www.xrepository.de/api/xrepository/urn:xoev-de:kosit:codeliste:icd_5:technischerBestandteilGenericode
- https://www.xrepository.de/api/xrepository/urn:xoev-de:kosit:codeliste:untdid.7161_3:technischerBestandteilGenericode
- https://www.xrepository.de/api/xrepository/urn:xoev-de:kosit:codeliste:untdid.7143_4:technischerBestandteilGenericode
- https://www.xrepository.de/api/xrepository/urn:xoev-de:kosit:codeliste:untdid.5305_3:technischerBestandteilGenericode
- https://www.xrepository.de/api/xrepository/urn:xoev-de:kosit:codeliste:untdid.5189_3:technischerBestandteilGenericode
- https://www.xrepository.de/api/xrepository/urn:xoev-de:xrechnung:codeliste:untdid.4461_3:technischerBestandteilGenericode
- https://www.xrepository.de/api/xrepository/urn:xoev-de:kosit:codeliste:untdid.4451_4:technischerBestandteilGenericode
- https://www.xrepository.de/api/xrepository/urn:xoev-de:kosit:codeliste:untdid.2005_4:technischerBestandteilGenericode
- https://www.xrepository.de/api/xrepository/urn:xoev-de:kosit:codeliste:untdid.1153_3:technischerBestandteilGenericode
- https://www.xrepository.de/api/xrepository/urn:xoev-de:kosit:codeliste:untdid.1001_4:technischerBestandteilGenericode
- https://www.xrepository.de/api/xrepository/urn:xoev-de:kosit:codeliste:currency-codes_3:technischerBestandteilGenericode
- https://www.xrepository.de/api/xrepository/urn:xoev-de:kosit:codeliste:country-codes_8:technischerBestandteilGenericode

File: erpnext/edi/doctype/common_code/common_code.py [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+	def before_validate(self):
+		if self.title and len(self.title) > 140:
+			self.title = self.title[:139] + "…"
+
+	def validate(self):
+		self.validate_distinct_references()
+
+	def validate_distinct_references(self):
+		"""Ensure no two Common Codes of the same Code List are linked to the same document."""
```
Comment (barredterra) [inline erpnext/edi/doctype/common_code/common_code.py:35]: For generating messages, it is very useful to have exactly one common code per ERPNext entity. However, for parsing messages, it would be useful to map multiple common codes to the same ERPNext entity. This is currently not possible due to this validation.

File: erpnext/edi/doctype/common_code/common_code.py [hunk 6]
```diff
@@ -28,6 +28,10 @@ class CommonCode(Document):
 		title: DF.Data
 	# end: auto-generated types
 
+	def before_validate(self):
+		if self.title and len(self.title) > 140:
+			self.title = self.title[:139] + "…"
+
```
Comment (blaggacao) [inline erpnext/edi/doctype/common_code/common_code.py]: If the title is used as canonical reference value (which I know I do), this might just start **randomly** braking some XML validations.

I'd construe this to be a view, not a data issue.

**Review by blaggacao — CHANGES_REQUESTED** [review]
I don't think manipulating origin reference data is a good idea. This may seem like a view, not a data problem.

File: erpnext/edi/doctype/common_code/common_code.py (hunk 6 above)
Comment (barredterra) [reply erpnext/edi/doctype/common_code/common_code.py]: What's the alternative? Our `Data` fieldtype doesn't support arbitrarily long data.

Comment (blaggacao) [reply erpnext/edi/doctype/common_code/common_code.py]: We'd have essentially two alternatives:

- Let it fail an the current field length (eventually with a better error message)
- Let it fail (very much less likely so) on an increased field length (eventually with a better error message), e.g. of `length: 300`

> Our `Data` fieldtype doesn't support arbitrarily long data.

We can dial it up, though.

Comment (blaggacao) [reply erpnext/edi/doctype/common_code/common_code.py]: We could instruct the user tho dial up the field length on failure via customization, for example.

Comment (barredterra) [reply erpnext/edi/doctype/common_code/common_code.py]: Increased the length, will leave the error message as is.

**Review by blaggacao — COMMENTED** [review]
Dismiss prior review :smile: I just need to test this a little more as a dep for `edi_template.py` and Colombian E-Invoice implementation. Getting there :hourglass_flowing_sand: 

Hm, that didn't work.... :thinking:

File: erpnext/edi/doctype/code_list/code_list.py [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+		self.db_set("default_common_code", None)
+
+		linked_docs = frappe.get_all(
+			"Common Code",
+			filters={"code_list": self.name},
+			fields=["name"],
+		)
+
+		for doc in linked_docs:
+			frappe.delete_doc("Common Code", doc.name, force=1)
```
Comment (barredterra) [inline erpnext/edi/doctype/code_list/code_list.py]:
```suggestion
			frappe.delete_doc("Common Code", doc.name)
```

@blaggacao was this intentional? If there are any existing links to such a force-deleted Common Code, the linking form will break.

File: erpnext/edi/doctype/common_code/common_code.py (hunk 2 above)
Comment (blaggacao) [reply erpnext/edi/doctype/common_code/common_code.py:107]: Checking now:

image

Comment (blaggacao) [reply erpnext/edi/doctype/common_code/common_code.py:107]: Freeze screen for about 30 sec:


image



That's not instant (as before), but probably ok.

Comment (blaggacao) [reply erpnext/edi/doctype/common_code/common_code.py:107]: Loading, otoh, was 8 sec. That's totally fine.

Comment (barredterra) [reply erpnext/edi/doctype/common_code/common_code.py:107]: > Freeze screen for about 30 sec:
> That's not instant (as before), but probably ok.

iirc, deletion was always slow. didn't change anything here.

**Comment by blaggacao** [comment]
Rebase done in order to prepare internal production backport which is hard to extract with a merge commit from develop.

File: erpnext/edi/doctype/code_list/code_list.py (hunk 7 above)
Comment (blaggacao) [reply erpnext/edi/doctype/code_list/code_list.py]: Hm, this looks like a dev-time QoL feature, in hindsight.

Deletion should almost never happen in normal operation.

File: erpnext/edi/doctype/common_code/common_code.py (hunk 5 above)
Comment (barredterra) [reply erpnext/edi/doctype/common_code/common_code.py:35]: @blaggacao any idea how to resolve this?

**Review by blaggacao — APPROVED** [review]
LGTM, I'm using this via branch https://github.com/barredterra/erpnext/pull/1 and it seems stable enough for this use case

Some of [these notes](https://github.com/frappe/erpnext/pull/43425#issuecomment-2386141316) might or might not serve for future evolution.

**Comment by blaggacao** [comment]
@barredterra This would need backports of https://github.com/frappe/frappe/pull/27954

File: erpnext/edi/doctype/common_code/common_code.py (hunk 5 above)
Comment (blaggacao) [reply erpnext/edi/doctype/common_code/common_code.py:35]: I'd recommend binding in the context of an EDI Template. Within this narrowed down context, this restriction is likely more suitable, but also the binding list is usually short enough that no duplicates would typically happen.

But this restriction, regardless, makes sense `on_update` within https://github.com/frappe/erpnext/pull/44174, specifically on [`bound_common_codes`](https://github.com/frappe/erpnext/pull/44174/files#diff-f3b34a58e2b8bf8d6892dfd41d9a5a4d076535000ad54d7ef12f669326128cd5R46)
